### ISSUE TITLE
How this library consume the Lookup/Write?

### ISSUE THREAD

**Issue by defaultbr** [cloudant/sync-android#547]
Hello, i have some questions how this library behave in some cases of replications

**In push case:**

1. Does it get one record per one from sqlite database and than do some post or it take a bunch of them and than send this bunch and handle the result? 
    
    if its in bunch, the consume of writes in second will be equal one or the server will separate each document from the bunch and count each one as one write?

2. every push replication is done after a lookup? Since you must know if the last local sequence is equal or above from the one at server, correctly?

**in pull case:**

1. Every pull replication will count as at least one Lookup since it need to check if local revision is equal or below the sequence from server? This is done at the changes by the `since sequence xxxx` and using limit = BATCH SIZE, correctly? 

2. This changes result return only the id or the full doc?

    in case of full docs should i consider it counting as just ONE lookup, correctly?

3. Otherwhise, if it only return the _ids since last sequence from device, the device will do new GET for each _id and get full doc at this time? (in this case i consider that each _id = 1 lookup)

     In this case all this is done in sequence, i mean, 1000 docs to pull, so organize the _ids list and do GET REQUEST(lookup) for each ?

4. Filtered pull replication change any of the above pull cases?

I'm trying to find the most effective way to handle multiple databases for multiple devices and some information can be shared in some cases and i know that i need to populate it from a external python server and this can rise the number of write/lookup **considerably** and i will regret to use per-user-database in the future

Thanks.

**Comment by ricellis** [comment]
Push case:

1. The records form the SQLite database are batched and sent to the `_bulk_docs` endpoint. The server will count one write request per document revision included in the batch sent. 
1. The replicator calculates the differences to the server using the `_revs_diff` endpoint, which is not counted as a lookup.

Pull case:

1. A request is made to `_changes` on the server to return the `_id`/`_rev` for changes since the last known sequence (from the previous replication or the previous batch of changes received in the current replication). Since this `_changes` does not count as a lookup if there are no changes to replicate then a pull replication will not even incur 1 lookup. Yes, the size of the changes fetched at a time is controlled by the `limit` parameter set via the [`changeLimitPerBatch` method](http://static.javadoc.io/com.cloudant/cloudant-sync-datastore-core/2.0.2/com/cloudant/sync/replication/ReplicatorBuilder.Pull.html#changeLimitPerBatch-int-).
1. The `_changes` requests do not include the docs and are *not* counted as lookups. 
1. The equivalent of `_revs_diff` is performed with the local database and then the missing document revisions are fetched from the server. You can consider it 1 lookup per missing `_rev` (for example if there are 3 revisions of the same `_id` on the server that are missing from the device there will need to be 3 lookups to fetch them all). In practice the fetch is done via the `_bulk_get` endpoint, but the server will count 1 lookup per document revision that is retrieved.
1. The filter is applied at the stage of getting `_changes` from the server, so anything rejected by the filter will not be fetched reducing the number of lookups required by the call to `_bulk_get`.

In summary:
* a push replication needs 1 write per document revision that needs to be written to the server.
* a pull replication needs 1 lookup per document revision that needs to be fetched to the device.

**Comment by defaultbr** [comment]
Awesome explanation, thank you so much
